src/ImageSharp/Processing/Processors/Quantization/IFrameQuantizer{TPixel}.cs

```diff
@@ -31,26 +31,23 @@ public interface IFrameQuantizer<TPixel> : IDisposable
         /// <returns>
         /// A <see cref="QuantizedFrame{TPixel}"/> representing a quantized version of the source frame pixels.
         /// </returns>
-        QuantizedFrame<TPixel> QuantizeFrame(
-            ImageFrame<TPixel> source,
-            Rectangle bounds);
+        QuantizedFrame<TPixel> QuantizeFrame(ImageFrame<TPixel> source, Rectangle bounds);
 
         /// <summary>
         /// Builds the quantized palette from the given image frame and bounds.
         /// </summary>
         /// <param name="source">The source image frame.</param>
         /// <param name="bounds">The region of interest bounds.</param>
-        /// <returns>The <see cref="ReadOnlyMemory{TPixel}"/> palette.</returns>
-        ReadOnlyMemory<TPixel> BuildPalette(ImageFrame<TPixel> source, Rectangle bounds);
+        /// <returns>The <see cref="ReadOnlySpan{TPixel}"/> palette.</returns>
+        ReadOnlySpan<TPixel> BuildPalette(ImageFrame<TPixel> source, Rectangle bounds);
```

Comment:
Can't find any related [guideline point](https://docs.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines), but I have a bad feeling returning a Span as a "result" of an operation, because the ownership of the backing memory is not clear from the API.

I think we should return `void` here, and have a separate `Palette` property instead.

Reply:
You've hit upon a point that led to me coding about 5 different version of this. 

I struggled with a `Palette` property for a couple of reasons:

1. The result is temporally coupled since it is not assigned on construction. The palette could be null (If using `ReadOnlyMemory<TPixel>`) or empty `ReadOnlySpan<TPixel>`
2. Using `ReadOnlyMemory<TPixel>` can lead to a larger palette length than actually required.

The `OctreeFrameQuantizer` doesn't return the final number of entries until you build the palette. You have to build it then slice the result reducing the length by the determined index count, returning `ReadOnlySpan` allowed this. 

https://github.com/SixLabors/ImageSharp/pull/1138/files#diff-2e327e0f312f1aae943b78b1145bd8adR90

If you don't do this you can end up with a larger Gif than you actually need. The `WuFrameQuantizer` is similar, but at least you know the length early on.

I'll give everything on last look over and see if I can improve things.

Reply:
> The palette could be null (If using `ReadOnlyMemory<TPixel>`) or empty `ReadOnlySpan<TPixel>`

We can also throw an `InvalidOperationException` if the palette has not yet been built.

> Using `ReadOnlyMemory<TPixel>` can lead to a larger palette length than actually required. [...] You have to build it then slice the result reducing the length by the determined index count, returning `ReadOnlySpan` allowed this. 

You can also slice a `ReadOnlyMemory<T>`.

This is no strong opinion however, we can also clarify the ownership model in the docs of `BuildPalette()`.

Reply:
> You can also slice a ReadOnlyMemory<T>.

Why oh why has that never occurred to me before?

Reply:
Quick question.... 

Would you consider is ok API wise for the QuantizedFrame to reference the FrameQuantizer palette rather than copying it? Currently I pass the span, copy it and use it for the lifetime of the frame.

Reply:
Scratch that. It leads to messy code in the Gif Encoder.